fs/fs_test.go

```diff
@@ -0,0 +1,158 @@
+	// Create blank in-memory FS
+	// One directory for source files, one for destination files
+	// One nested directory for source files being copied with structure
+	// One file in the root directory to copy into the destination
+	// One empty directory, one with no read permissions
+	tempFS := afero.NewMemMapFs()
+	_ = tempFS.Mkdir("src", 0755)
+	_, _ = tempFS.Create("src/srcfile1.go")
+	_ = tempFS.Mkdir("dest", 0755)
+	_ = tempFS.MkdirAll("nestdir/levelone", 0755)
+	_, _ = tempFS.Create("nestdir/levelone/srcfile2.go")
+	_, _ = tempFS.Create("srcfile3.go")
+	_ = tempFS.Mkdir("empty", 0755)
+	_ = tempFS.Mkdir("restricted", 0000)
+	_, _ = tempFS.Create("restricted/cantread.go")
+
+	type args struct {
+		targetFs afero.Fs
+		src      string
+		dest     string
+		fileOnly bool
+	}
+	tests := []struct {
+		name    string
+		args    args
+		wantErr bool
+	}{
+		// Add test entries here
+		{"Copy from src to dest dir", args{tempFS, "src", "dest", true}, false},
+		{"Copy with source structure", args{tempFS, "nestdir", "dest", false}, false},
+		{"Copy file from parent of dest", args{tempFS, "srcfile3.go", "dest", false}, false},
+		{"Destination directory does not exist", args{tempFS, "nestdir", "newdest", false}, false},
+		{"Source directory is empty", args{tempFS, "empty", "dest", false}, false},
+		{"Source file does not exist", args{tempFS, "nofile.go", "dest", true}, false},
+		{"Source directory has no read perms", args{tempFS, "restricted", "dest", false}, false},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			if err := CopyFromOS(tt.args.targetFs, tt.args.src, tt.args.dest, tt.args.fileOnly); (err != nil) != tt.wantErr {
+				t.Errorf("CopyFromOS() error = %v, wantErr %v", err, tt.wantErr)
+			}
+		})
+	}
+}
+
+func TestStreamFiles(t *testing.T) {
+	// Filters:
+	noFilter := []func(string) bool{}
+	markFilter := []func(string) bool{MarkdownOnly}
+	underscoreFilter := []func(string) bool{NoUnderscores}
+	allFilter := []func(string) bool{MarkdownOnly, NoUnderscores}
+
+	type args struct {
+		path    string
+		files   chan string
+		filters []func(file string) bool
+	}
+	tests := []struct {
+		name    string
+		args    args
+		wantErr bool
+	}{
+		// Add test entries here
+		{"No filters", args{"../example", make(chan string), noFilter}, false},
+		{"Markdown filters", args{"../example", make(chan string), markFilter}, false},
+		{"Underscore filters", args{"../example", make(chan string), underscoreFilter}, false},
+		{"All filters", args{"../example", make(chan string), allFilter}, false},
+		{"Path does not exist", args{"notexist", make(chan string), allFilter}, false},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			channel := tt.args.files
+			go func() {
+				// Drain all data that goes into channel
+				for i := range channel {
+					_ = i
+				}
+			}()
+			if err := StreamFiles(tt.args.path, channel, tt.args.filters...); (err != nil) != tt.wantErr {
```

Comment:
`(err != nil) != tt.wantErr` Nice 😄 Depending on your personal preference, you may also write `err != nil && !tt.wantErr` here.

Reply:
Honestly, that snippet was taken directly from gotests :P I thought it was nice (not the most readable) and so left it like that. 

About the other option `err != nil && !tt.wantErr`, wouldn't it cause an issue when `tt.wantErr` is true and the function was supposed to return an error?

Reply:
> About the other option err != nil && !tt.wantErr, wouldn't it cause an issue when tt.wantErr is true and the function was supposed to return an error?

Absolutely. I managed to forget about that :-)